crates/turborepo-lib/src/shim.rs

```diff
@@ -176,50 +178,234 @@ struct PackageJson {
+#[serde(rename_all = "camelCase")]
+struct YarnRc {
+    pnp_unplugged_folder: PathBuf,
 }
 
-impl LocalTurboState {
-    pub fn infer(repo_root: &Path) -> Option<Self> {
-        let local_turbo_path = repo_root.join("node_modules").join(".bin").join({
+impl Default for YarnRc {
+    fn default() -> Self {
+        Self {
+            pnp_unplugged_folder: [".yarn", "unplugged"].iter().collect(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TurboState {
+    bin_path: Option<PathBuf>,
+    version: &'static str,
+    repo_state: Option<RepoState>,
+}
+
+impl Default for TurboState {
+    fn default() -> Self {
+        Self {
+            bin_path: env::current_exe().ok(),
+            version: get_version(),
+            repo_state: None,
+        }
+    }
+}
+
+impl TurboState {
+    pub fn platform_package_name() -> &'static str {
+        const ARCH: &str = {
+            #[cfg(target_arch = "x86_64")]
+            {
+                "64"
+            }
+            #[cfg(target_arch = "aarch64")]
+            {
+                "arm64"
+            }
+            #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+            {
+                "unknown"
+            }
+        };
+
+        const OS: &str = {
+            #[cfg(target_os = "macos")]
+            {
+                "darwin"
+            }
+            #[cfg(target_os = "windows")]
+            {
+                "windows"
+            }
+            #[cfg(target_os = "linux")]
+            {
+                "linux"
+            }
+            #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
+            {
+                "unknown"
+            }
+        };
+
+        formatcp!("turbo-{}-{}", OS, ARCH)
+    }
+
+    pub fn binary_name() -> &'static str {
+        {
             #[cfg(windows)]
             {
-                "turbo.cmd"
+                "turbo.exe"
             }
             #[cfg(not(windows))]
             {
                 "turbo"
             }
-        });
-
-        if !local_turbo_path.exists() {
-            debug!(
-                "No local turbo binary found at: {}",
-                local_turbo_path.display()
-            );
-            return None;
         }
+    }
+
+    #[allow(dead_code)]
+    pub fn version() -> &'static str {
+        include_str!("../../../version.txt")
+            .lines()
+            .next()
+            .expect("Failed to read version from version.txt")
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct LocalTurboState {
+    bin_path: PathBuf,
+    version: String,
+}
+
+impl LocalTurboState {
+    // Hoisted strategy:
+    // - `npm install`
+    // - `yarn`
+    // - `yarn install --flat`
+    // - berry (nodeLinker: "node-modules")
+    //
+    // This also supports people directly depending upon the platform version.
+    fn generate_hoisted_path(root_path: &Path) -> Option<PathBuf> {
```

Comment:
Would these functions benefit from our newly-acquired path library?

Reply:
I _very_ carefully didn't type a single `\` or `/` because I didn't have good tools for this. Right now it's assuredly correct since they're all relative path appends to a `root_path` value calculated by Rust, but yeah; that's a further refactoring for after this lands.

I really want something like https://github.com/zertosh/path_macro except better that does all of these "magic paths" into `&'static str`, but that's hard to justify building.

Something like:
`let path = smart_path_macro!(root_path / "node_modules" / "turbo" / "node_modules")`

Would validate that all static segments are good and then spit out:
```
#[cfg(not(windows))]
let path = root_path.join("node_modules/turbo/node_modules")
#[cfg(windows)]
let path = root_path.join("node_modules\\turbo\\node_modules")
```

I'll catch this on the next refactor PR.

Reply:
Yeah I'd consider taking a look at the path library and seeing if it fits here. If it doesn't, then perhaps we can open linear tickets for features and add them.

Reply:
The existing `AbsolutePath` type can be either Windows or Unix. The only thing it validates is `Path.isAbsolute` at construction. I don't actually care about that at all in this context, all paths are constructed explicitly using `join()`.

In the current state it provides me zero value, as demonstrated by the number of issues that we had with the git stuff on Windows.

If you look at the implementation:
```
    pub fn join<P: AsRef<Path>>(&self, other: P) -> AbsolutePathBuf {
        let path = self.0.join(other);
        assert!(path.is_absolute());
        AbsolutePathBuf(path)
    }
```

The baseline rules of `join` guarantee that assertion in advance:
https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html#method.push

I appreciate that we want to push this forward, but that's a task in and of itself, not something to attach as a rider this late. (The lib landed after this work had reached "mostly done" and is only in tree because of a rebase.)